drivers/adc/adc_mcux_adc16.c

```diff
@@ -19,19 +24,79 @@ LOG_MODULE_REGISTER(adc_mcux_adc16);
 
 struct mcux_adc16_config {
 	ADC_Type *base;
+#ifndef CONFIG_ADC16_ENABLE_EDMA
 	void (*irq_config_func)(const struct device *dev);
+#endif
 	bool channel_mux_b;
+	uint32_t clk_source;
+	uint32_t long_sample;
+	uint32_t hw_trigger_src;
+	uint32_t dma_channel;
+	uint32_t dma_slot;
+	bool nxp_high_speed;
+	bool nxp_continues_convert;
 };
 
+#ifdef CONFIG_ADC16_ENABLE_EDMA
+struct adc_edma_config {
+	const char *dma_name;
+	int32_t state;
+	uint32_t dma_channel;
+	void (*irq_call_back)(void);
+	struct dma_config dma_cfg;
+	struct dma_block_config dma_block;
+};
+#endif
+
 struct mcux_adc16_data {
 	const struct device *dev;
 	struct adc_context ctx;
+#ifdef CONFIG_ADC16_ENABLE_EDMA
+	const struct device *dev_dma;
+	struct adc_edma_config adc_dma_config;
+#endif
 	uint16_t *buffer;
 	uint16_t *repeat_buffer;
 	uint32_t channels;
 	uint8_t channel_id;
 };
 
+#define DEV_CFG(dev) ((const struct mcux_adc16_config *const)dev->config)
+#define DEV_DATA(dev) ((struct mcux_adc16_data *)dev->data)
+#define DEV_BASE(dev) ((ADC_Type *)DEV_CFG(dev)->base)
+
+#ifdef CONFIG_ADC16_ENABLE_EDMA
+static void adc_dma_callback(const struct device *dma_dev, void *callback_arg,
+			     uint32_t channel, int error_code)
+{
+	struct device *dev = (struct device *)callback_arg;
+	struct mcux_adc16_data *data = DEV_DATA(dev);
+
+	LOG_DBG("DMA done");
+	adc_context_on_sampling_done(&data->ctx, dev);
+}
+#endif
+
+#ifdef CONFIG_ADC_HW_TRIGGER
+static void adc_hw_trigger_enable(const struct device *dev)
+{
+	const struct mcux_adc16_config *config = dev->config;
+	/* enable ADC trigger channel */
+	if (DEV_BASE(dev) == (ADC_Type *)DT_INST_REG_ADDR(0)) {
```

Comment:
The comparisons rely on devicetree instances corresponding to the device instances found in the SoC. That is not guaranteed. You will need another way of determining the register flags to be set (perhaps some preprocessor magic for storing them in `dev->config`?)

Reply:
ok, I add the instance magic code to compare.